Source/Blazorise/Components/Validation/Validation.razor.cs

```diff
@@ -110,7 +119,18 @@ internal async Task InitializeInput( IValidationInput inputComponent )
         initialized = true;
     }
 
-    internal async Task InitializeInputPattern<T>( string patternString, T value )
+    /// <summary>
+    /// Initializes or updates the input validation pattern. If the pattern string changes,
+    /// a new regex pattern is created, and the input is re-validated if applicable.
+    /// </summary>
+    /// <typeparam name="T">The type of the value to validate against the pattern.</typeparam>
+    /// <param name="patternString">The regex pattern string for validation.</param>
+    /// <param name="value">The current value of the input to validate against the new pattern.</param>
+    /// <remarks>
+    /// This method optimizes performance by avoiding re-instantiation of the regex pattern if it has not changed.
+    /// It ensures that validation is only re-triggered if the component has been initialized and validation conditions are met.
+    /// </remarks>
+    public async Task InitializeInputPattern<T>( string patternString, T value )
```

Comment:
I don't know much about our validation works internally. But do we need to expose this and `InitializeInputExpression`? Can't we just expose a single Initialization that handles accordingly to whatever is defined?

Since both methods will only really do something if the `Validation` is configured accordingly, seems like an internal implementation detail that should be part of a more generic **Initialization** method?

Reply:
I don't remember TBH. I think some components have different initialization flow so I made them as separate function to be more flexible. Maybe we can indeed optimize it by having only one function.

Reply:
Yea otherwise you're making it so the consumer has to study how he's going to initialize the Validation.
Try to put yourself on his shoes. How would you use this api?
Call every Initialize in a row? Why do I explicitly need to initialize the other ones? Does it need to happen somewhere else?
```
InitializeInput();
InitializeInputPattern();
InitializeInputExpression();
```
Will it work? Do I need additional checks? The moment you make a user have to think too much or investigate the internal api, then the exposed api might not be clear.

Reply:
Its might be difficult. See this

**TimeEdit**;

```cs
if ( parameters.TryGetValue<Expression<Func<TValue>>>( nameof( TimeExpression ), out var expression ) )
    await ParentValidation.InitializeInputExpression( expression );

if ( parameters.TryGetValue<string>( nameof( Pattern ), out var pattern ) )
{
    // make sure we get the newest value
    var value = parameters.TryGetValue<TValue>( nameof( Time ), out var inTime )
        ? inTime
        : InternalValue;

    await ParentValidation.InitializeInputPattern( pattern, value );
}

await InitializeValidation();
```

**TextEdit**:

```cs
if ( parameters.TryGetValue<Expression<Func<string>>>( nameof( TextExpression ), out var expression ) )
    await ParentValidation.InitializeInputExpression( expression );

if ( parameters.TryGetValue<string>( nameof( Pattern ), out var pattern ) )
{
    // make sure we get the newest value
    var value = parameters.TryGetValue<string>( nameof( Text ), out var paramText )
        ? paramText
        : InternalValue;

    await ParentValidation.InitializeInputPattern( pattern, value );
}

await InitializeValidation();
```

We have different parameter names for parameters and different data types. Some are generic(TValue), some could be known, like `string`.

Reply:
Right. I don't like it at all.
Anyway, could **parameters** be wrapped in a dedicated method and we abstract away this bit?
Like 
`OnParametersSetValidation(parameters);`

Otherwise we can just take the approach of properly documenting the process of creating a validatable blazorise component with these nuances and refer anyone to it.

Reply:
Also, not all components have Pattern validation. So that's that.

Reply:
What do you think of this approach?

```cs
public async Task Initialize<TValue>( IValidationInput inputComponent, ParameterView parameters, string valueParameterName, string expressionParameterName )
{
    if ( parameters.TryGetValue<Expression<Func<TValue>>>( expressionParameterName, out var paramExpression ) )
        await InitializeInputExpression( paramExpression );

    if ( parameters.TryGetValue<string>( nameof( Pattern ), out var paramPattern ) )
    {
        // make sure we get the newest value
        var value = parameters.TryGetValue<TValue>( valueParameterName, out var paramValue )
            ? paramValue
            : inputComponent.ValidationValue;

        await InitializeInputPattern( paramPattern, value );
    }

    await InitializeInput( inputComponent );
}
```

Reply:
Check it here: https://github.com/Megabit/Blazorise/commit/be5a6e51687dce015004dc86bf1dce8d0c3b7f91

If you think it is worth it I will merge it here in this PR. At least we have less code to intitialise.

Reply:
That seems way better to me. The fact that the parameters argument is in there already hints at the consumer this should be in the ParametersSet. But maybe the method name should be more descriptive somehow? Because altought it's called Initialize, it will be called multiple times.
Initialize generally is something that is called once.

Reply:
Oh there's still quite a bit of ceremony in fact.... hmm

Reply:
If it's only called once = > validationInitialized why do we need it to be on the parameters set again? Because we need to access the parameters collection?

Reply:
Yes, we need to get the latest value from the parameters. Which I now realized the new code might be wrong. Previously I checked Expression and Pattern parameters on each component change. And if I leave that code placed in the InitializeValidation it will not work. Possibly breaking users code if they change those parameters dynamically.

Back to the drawing board.

Reply:
OK I fixed it. Now it should work as expected.

And renamed `Initialize` into `InitializeValidationParameters`.

Reply:
https://github.com/Megabit/Blazorise/commit/0ca72d7f5ddc07dbb5efee426540bf118566296f

Reply:
In the end, I think it is best to have separate APIs for each validation process. While it works to have them all under the one API I think it is not that flexible for us and for users. Going to merge this PR now.